### PR TITLE
Migrate to V1 Engine w/ Optimum-based

### PR DESCRIPTION
**Pull request by rebel-eunji**
#### Block Table and Scheduler
vllm_v1-kv-cache-2

- I’ve tried to keep the modifications minimal.
- It simply uses VLLM’s original block table and KV cache manager.
- The scheduling process and priority is changed for using optimum-rbln.

#### Worker and Runner
vllm_v1-worker-runner-2

- Additionally, I reused `ModelInputForRBLN` class for compatibility between V0 and V1 with some modification.

### 📌 Related Issues / Tickets

* Related to #8 

---

### ✅ Changes

* [x] Feature
* [ ] Bug Fix
* [ ] Documentation
* [ ] Refactor
* [ ] Optimization
* [ ] Other (please describe):

---

### 📸 Screenshots / Logs (if applicable)

---

### 🧪 How to Test

1. Run [vLLM official example code](https://github.com/vllm-project/vllm/blob/main/examples/offline_inference/llm_engine_example.py)
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📋 Checklist

* [ ] I’ve added unit tests or updated existing ones.
* [ ] I’ve updated relevant documentation.
* [ ] I’ve checked compatibility (e.g., hardware/backend/platform).

---

### 💬 Notes
- This PR only works for decoder-only models.
- AsyncLLMEngine, Multimodal, and PP is not supported yet. The features will be supported soon.

### CODE REVIEW

**Comment by rebel-eunji** [comment]
CI is currently failing because of a version mismatch in `optimum-rbln`. This issue is not caused by this PR.

**Comment by rebel-jonghewk** [comment]
@rebel-eunji Can you enable V1 so that the CI runs with V1 enabled?

File: vllm_rbln/v1/worker/optimum_worker.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    @torch.inference_mode()
+    def determine_available_memory(self) -> int:
+        max_model_len = self.vllm_config.model_config.max_model_len
+        kv_cache_spec = self.model_runner.get_kv_cache_spec()
+
+        num_gpu_blocks = self.scheduler_config.max_num_seqs + 1
+        block_size = self.cache_config.block_size
+        return num_gpu_blocks * block_size * len(
+            kv_cache_spec.items()) * max_model_len
```
Comment (rebel-jonghewk) [inline vllm_rbln/v1/worker/optimum_worker.py:74]: I think this should be revisited.

Comment (rebel-eunji) [reply vllm_rbln/v1/worker/optimum_worker.py]: @rebel-jonghewk  Yes, it was incorrect. vLLM uses this function to calculate the available GPU blocks. ([function](https://github.com/vllm-project/vllm/blob/b6553be1bc75f046b00046a4ad7576364d03c835/vllm/v1/core/kv_cache_utils.py#L644)).
So I modified that formula to calculate the number of blocks we need.

num_blocks = int(available_memory // page_size // num_layers)
-> available_memory  = num_blocks * page_size * num_layers

File: examples/optimum/run_decoder_only.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ async def generate(engine: AsyncLLMEngine, prompt: str, model: str,
     example_input = {
         "stream": True,
         "temperature": 0.0,
-        "request_id": requst_id,
+        "request_id": str(requst_id),
```
[the remark this answers is not in the document]
Comment (rebel-eunji) [reply examples/optimum/run_decoder_only.py:30]: Yes Thanks. I will fix with the parameter name. 👍

File: .github/workflows/rbln_optimum_ci.yaml [hunk 3]
```diff
@@ -79,6 +79,74 @@ jobs:
         run: |
           pip install --force-reinstall --no-cache-dir dist/vllm_rbln*.whl
 
+      - name: Run decoder-only test (eager attn) (V1)
+        env:
+          VLLM_LOGGING_LEVEL: DEBUG
+        run: >
+          VLLM_USE_V1=1 python3 examples/optimum/run_decoder_only.py
```
Comment (rebel-jonghewk) [inline .github/workflows/rbln_optimum_ci.yaml:86]: Isn't `VLLM_USE_V1` true as default?

**Comment by rebel-jonghewk** [comment]
> @rebel-jonghewk To run V1 in CI, we need to set like below
> 
> 1. Set the environment variable `VLLM_USE_V1=1` when run the vllm code
> 2. V1 doesn't support encoder-decoder models and encoder models. So we need to except them.

What happens when user's use V1 with encoder-decoder & encoder models? Does it automatically fallback to V0? (If not it should be.)

File: .github/workflows/rbln_optimum_ci.yaml [hunk 4]
```diff
@@ -79,6 +79,74 @@ jobs:
...
+          --num_input_prompt ${{ env.NUM_INPUT_PROMPT }}
+          --model_id ${{ env.REBEL_VLLM_PRE_COMPILED_DIR }}/llama2-7b_batch2
+          --prompt_txt ${{ env.REBEL_VLLM_PRE_COMPILED_DIR }}/prompts/copy_prompts.txt
+          --golden_json ${{ env.REBEL_VLLM_PRE_COMPILED_DIR }}/golden/golden_llama7b_result_copy_prompts.json
+
+      - name: Run decoder-only test (Flash-attention mode) (V1)
+        env:
+          VLLM_LOGGING_LEVEL: DEBUG
```
Comment (rebel-jonghewk) [inline .github/workflows/rbln_optimum_ci.yaml:94]: How about add `VLLM_LOGGING_LEVEL: DEBUG` below `RBLN_VERBOSE: 6`?

File: vllm_rbln/v1/core/optimum_scheduler.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+        scheduled_timestamp = time.monotonic()
+
+        # NOTE The scheduling process is changed like below.
+        # (1) vllm-rbln distinguishes
+        #   between requests in the prefill and decode phases.
+        #   If a request is in the prefill phase,
+        #   it is given priority and processed exclusively (only one at a time).
+        # (2) For (1), vllm-rbln schedules the requests WAITING -> RUNNING.
+        #   In the vLLM, requests are scheduled RUNNING -> WAITING.
+
```
Comment (rebel-jonghewk) [inline vllm_rbln/v1/core/optimum_scheduler.py:104]: Just to confirm my understanding of the logic—the differences from the original v1 implementation are:

Reordering the execution of running and waiting requests

Adding a break condition for req_index > 0

Disabling LoRA

Are there any other changes I might have missed?

File: vllm_rbln/v1/worker/optimum_model_runner.py [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        self,
+        scheduler_output: "SchedulerOutput",
+    ) -> ModelInputForRBLN:
+        """
+        :return: tuple[
+            attn_metadata: layer-to-attention_metadata mapping,
+            attention_cuda_graphs: whether attention can run in cudagraph
+            logits_indices, spec_decode_metadata
+        ]
+        """
```
Comment (rebel-jonghewk) [inline vllm_rbln/v1/worker/optimum_model_runner.py:245]: This should be fixed.

File: .github/workflows/rbln_optimum_ci.yaml (hunk 3 above)
Comment (rebel-eunji) [reply .github/workflows/rbln_optimum_ci.yaml:86]: Yes.
https://github.com/vllm-project/vllm/blob/e7b204268132cb775c139574c1ff4ad7e15c8f66/vllm/engine/arg_utils.py#L1026C1-L1042C1
We need to export the envs to use V1.

File: vllm_rbln/v1/core/optimum_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+        # With V0 -> V1, block_manager is changed to kv_cache_manager
+        self.kv_cache_manager = KVCacheManager(
+            kv_cache_config=kv_cache_config,
+            max_model_len=self.max_model_len,
+            enable_caching=self.cache_config.enable_prefix_caching,
+            caching_hash_algo=self.cache_config.prefix_caching_hash_algo,
+            use_eagle=self.use_eagle,
+            log_stats=self.log_stats,
+            enable_kv_cache_events=self.enable_kv_cache_events,
+        )
```
[the remark this answers is not in the document]
Comment (rebel-eunji) [reply vllm_rbln/v1/core/optimum_scheduler.py]: Great point! I removed this init. 👍

File: vllm_rbln/v1/core/optimum_scheduler.py (hunk 5 above)
Comment (rebel-eunji) [reply vllm_rbln/v1/core/optimum_scheduler.py:75]: No, That's all.

File: vllm_rbln/v1/worker/optimum_model_runner.py [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+        total_num_scheduled_tokens = scheduler_output.total_num_scheduled_tokens
+        assert total_num_scheduled_tokens > 0
+        num_reqs = self.input_batch.num_reqs
+        assert num_reqs > 0
+
+        # OPTIMIZATION: Start copying the block table first.
+        # This way, we can overlap the copy with the following CPU operations.
+        self.input_batch.block_table.commit(num_reqs)
```
[the remark this answers is not in the document]
Comment (rebel-eunji) [reply vllm_rbln/v1/worker/optimum_model_runner.py:254]: Yes. We don't need commit (cpu -> gpu tensor copy). I'll fix it to use block_table in `cpu` without using commit.

File: vllm_rbln/platform.py [hunk 9]
```diff
@@ -129,10 +134,26 @@ def check_and_update_config(cls, vllm_config: VllmConfig) -> None:
...
+            is_pooling = task in pooling_tasks
+            is_generate = task == "generate"
+
+        if is_pooling and envs.VLLM_USE_V1:
+            raise ValueError("Pooling models are only supported on v0.")
+
+        if is_generate and cls.supports_v1(
+                model_config
+        ) and not envs.VLLM_USE_V1 and not model_config.is_encoder_decoder:
+            logger.warning("V0 support for decoder models is deprecated.")
```
Comment (rebel-jonghewk) [inline vllm_rbln/platform.py:155]: @rebel-eunji I've changed this to a warning, as there may still be valid cases where V0 can be used for now.

File: vllm_rbln/v1/worker/optimum_model_runner.py [hunk 10]
```diff
@@ -268,10 +268,12 @@ def _prepare_inputs(
         if num_prefill_reqs > 1 or (num_prefill_reqs >= 1
                                     and num_decode_reqs > 0):
             raise RuntimeError(
-                "prefill stage request cannot processed with other requests.")
+                "Prefill stage request cannot processed with other requests.")
 
         if num_prefill_reqs > 0:
             is_prefill = True
+        elif num_decode_reqs == 1 and scheduler_output.scheduled_cached_reqs[0].resumed_from_preemption == True:
```
Comment (rebel-jonghewk) [inline vllm_rbln/v1/worker/optimum_model_runner.py]: @rebel-eunji why do you need to add `num_decode_reqe == 1` ?

**Review by rebel-jonghewk — APPROVED** [review]
LGTM!
